### PR TITLE
#222 duplicate event feature 

### PR DESCRIPTION
**Pull request by rezk2ll**
**DEMOS**:

create a duplicate event as an organizer: https://streamable.com/evfnyg
duplicate an event as a non-organizer: https://streamable.com/ozqd68

### CODE REVIEW

File: src/esn.calendar.libs/app/app.constants.js [hunk 1]
```diff
@@ -332,6 +332,11 @@
      */
     .constant('CAL_ALARM_MODIFY_COMPARE_KEYS', ['action', 'attendee', 'description', 'summary', 'trigger'])
 
+    /**
+     * When duplicating an event, the JSON keys on the event will be copied.
+     */
+    .constant('CAL_EVENT_DUPLICATE_KEYS', ['attendees', 'title', 'start', 'end', 'location', 'description', 'alarm', 'xOpenpaasVideoconference'])
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/app.constants.js:338]: What about resources and privacy?

**Comment by huy-ta** [comment]
The second commit's message is too short imo, and maybe you don't even need that commit because there should be one commit.

**Comment by rezk2ll** [comment]
> The second commit's message is too short imo, and maybe you don't even need that commit because there should be one commit.

commit message is longer now :trollface: 
i'd like to keep the actual code and test in separate commits.

File: src/esn.calendar.libs/app/app.constants.js (hunk 1 above)
Comment (rezk2ll) [reply src/esn.calendar.libs/app/app.constants.js]: thanks for spotting that, we copy the `privacy` now.

for the `resources` AFAIK they are not actually bound to the event object.

File: src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js [hunk 2]
```diff
@@ -2090,5 +2098,89 @@ describe('The CalEventFormController controller', function() {
...
+
+        expect(scope.cancel).to.have.been.called;
+      });
+
+      it('should open a new event form', function() {
+        scope.duplicateEvent();
+        $timeout.flush();
+
+        expect(calOpenEventFormMock).to.have.been.called;
+      });
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js:2171]: I think these two test cases can be merged into one, just "should close the previous event form and open a new event form after a delay".

File: src/esn.calendar.libs/app/app.constants.js [hunk 3]
```diff
@@ -332,6 +332,11 @@
      */
     .constant('CAL_ALARM_MODIFY_COMPARE_KEYS', ['action', 'attendee', 'description', 'summary', 'trigger'])
 
+    /**
+     * When duplicating an event, the JSON keys on the event will be copied.
+     */
+    .constant('CAL_EVENT_DUPLICATE_KEYS', ['attendees', 'title', 'start', 'end', 'location', 'description', 'xOpenpaasVideoconference', 'class'])
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/app.constants.js:338]: What about resources and alarm?

File: src/esn.calendar.libs/app/components/event/form/event-form.controller.js [hunk 4]
```diff
@@ -588,4 +594,62 @@ function CalEventFormController(
...
+    return CalendarShell.fromIncompleteShell(details);
+  }
+
+  function _generateVideoConferenceUrl() {
+
+    return VideoConfConfigurationService.getOpenPaasVideoconferenceAppUrl()
+      .then(openPaasVideoconferenceAppUrl => `${openPaasVideoconferenceAppUrl}${uuid4.generate()}`)
+      .catch(err => {
+        $log.error('Cannot generate a new video conference URL', err);
+        _displayNotification(notificationFactory.weakError, null, 'Failed to create a new video conference room');
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/components/event/form/event-form.controller.js:646]: I guess this already has a translation, right?

File: src/esn.calendar.libs/app/app.constants.js (hunk 3 above)
Comment (rezk2ll) [reply src/esn.calendar.libs/app/app.constants.js:338]: believe it or not, a resource is actually an attendee:
image

File: src/esn.calendar.libs/app/components/event/form/event-form.controller.js (hunk 4 above)
Comment (rezk2ll) [reply src/esn.calendar.libs/app/components/event/form/event-form.controller.js:646]: nope, added now

File: src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js (hunk 2 above)
Comment (rezk2ll) [reply src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js]: done

File: src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js [hunk 5]
```diff
@@ -2090,5 +2098,84 @@ describe('The CalEventFormController controller', function() {
...
+
+      it('should not generate a new video conference link if the original event didn\'t have one', function() {
+        scope.editedEvent = CalendarShell.fromIncompleteShell(eventSkeleton);
+
+        scope.duplicateEvent();
+
+        expect(VideoConfConfigurationServiceMock.getOpenPaasVideoconferenceAppUrl).to.not.have.been.called;
+      });
+
+      it('should close the previous event form and open a new event form abter a brief delday', function() {
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js:2159]:
```suggestion
      it('should close the previous event form and open a new event form after a brief delay', function() {
```

:trollface:

**Review by renaudboyer — APPROVED** [review]

**Review by asmaNode — APPROVED** [review]

**Review by huy-ta — APPROVED** [review]

**Pull request merged**